lib/spack/spack/test/spec_set.py

```diff
@@ -0,0 +1,298 @@
+                    'gmake@4.0',
+                    'appres@1.0.4',
+                    'allinea-reports@6.0.4'
+                ]},
+            ]
+        }
+    }
+    spec_set = CombinatorialSpecSet(yaml_file, False)
+    specs = list(spec for spec in spec_set)
+    assert len(specs) == 3
+
+
+def test_spec_set_with_specs():
+    """Make sure we only see the specs mentioned in the include"""
+    yaml_file = {
+        'spec-set': {
+            'include': ['gmake', 'appres'],
+            'matrix': [
+                {'specs': [
+                    'gmake@4.0',
+                    'appres@1.0.4',
+                    'allinea-reports@6.0.4'
+                ]},
+                {'compilers': {
+                    'gcc': {
+                        'versions': ['4.2.1', '6.3.0']
+                    }, 'clang': {
+                        'versions': ['8.0', '3.8']
+                    }
+                }},
+            ]
+        }
+    }
+    spec_set = CombinatorialSpecSet(yaml_file, False)
+    specs = list(spec for spec in spec_set)
+    assert len(specs) == 8
+
+
+def test_spec_set_packages_no_matrix():
+    """The matrix property is required, make sure we error out if it is
+    missing"""
+    yaml_file = {
+        'spec-set': {
+            'include': ['gmake'],
+            'packages': {
+                'gmake': {
+                    'versions': ['4.0']
+                },
+                'appres': {
+                    'versions': ['1.0.4']
+                },
+                'allinea-reports': {
+                    'versions': ['6.0.4']
+                }
+            },
+        }
+    }
+    with pytest.raises(ValidationError):
+        CombinatorialSpecSet(yaml_file)
+
+
+def test_spec_set_get_cdash_array():
+    """Make sure we can handle multiple cdash sites in a list"""
+    yaml_file = {
+        'spec-set': {
+            'cdash': ['http://example.com/cdash', 'http://example.com/cdash2'],
+            'project': 'testproj',
+            'matrix': [
+                {'packages': {
+                    'gmake': {'versions': ['4.0']},
+                }},
+                {'compilers': {
+                    'gcc': {'versions': ['4.2.1', '6.3.0']},
+                    'clang': {'versions': ['8.0', '3.8']},
+                }},
+            ]
+        }
+    }
+
+    spec_set = CombinatorialSpecSet(yaml_file)
+    assert spec_set.cdash == [
+        'http://example.com/cdash', 'http://example.com/cdash2']
+    assert spec_set.project == 'testproj'
+
+
+def test_compiler_specs():
+    spec_set = CombinatorialSpecSet(basic_yaml_file, False)
+    compilers = spec_set._compiler_specs({
+        'gcc': {
+            'versions': ['4.2.1', '6.3.0']
+        }, 'clang': {
+            'versions': ['8.0', '3.8']
+        }})
+
+    assert Spec('%gcc@4.2.1') in compilers
+    assert Spec('%gcc@6.3.0') in compilers
+    assert Spec('%clang@8.0') in compilers
+    assert Spec('%clang@3.8') in compilers
```

Comment:
perhaps also assert that `len(compilers) == 4`

Reply:
Added that check.